File: ios/LeanIOS/Ads/AdMobAdsManager.swift
#if AD_NETWORK_ADMOB
import Foundation
import UIKit
import GoogleMobileAds

final class AdMobAdsManager: NSObject, AdsManaging, FullScreenContentDelegate {
    private var config = AdConfig(dict: [:])
    private var interstitial: InterstitialAd?
    private var rewarded: RewardedAd?
    private var appOpen: AppOpenAd?
    private var rewardCallback: ((String, Int) -> Void)?

    // Google-provided test unit IDs.
    private let testInterstitial = "ca-app-pub-3940256099942544/4411468910"
    private let testRewarded = "ca-app-pub-3940256099942544/1712485313"
    private let testAppOpen = "ca-app-pub-3940256099942544/5575463023"

    func start(with config: AdConfig) {
        self.config = config
        MobileAds.shared.start { _ in
            self.loadInterstitial(); self.loadRewarded(); self.loadAppOpen()
        }
    }

    private func iid(_ v: String, _ fallback: String) -> String { v.isEmpty ? fallback : v }

    private func loadInterstitial() {
        InterstitialAd.load(with: iid(config.admobInterstitial, testInterstitial), request: Request()) { ad, err in
            if let err = err { NSLog("[AdMob] interstitial load err: \(err)"); return }
            ad?.fullScreenContentDelegate = self; self.interstitial = ad
        }
    }

    private func loadRewarded() {
        RewardedAd.load(with: iid(config.admobRewarded, testRewarded), request: Request()) { ad, err in
            if let err = err { NSLog("[AdMob] rewarded load err: \(err)"); return }
            ad?.fullScreenContentDelegate = self; self.rewarded = ad
        }
    }

    private func loadAppOpen() {
        AppOpenAd.load(with: iid(config.admobAppOpen, testAppOpen), request: Request()) { ad, err in
            if let err = err { NSLog("[AdMob] appOpen load err: \(err)"); return }
            ad?.fullScreenContentDelegate = self; self.appOpen = ad
        }
    }

    func showInterstitial(from vc: UIViewController) {
        if let ad = interstitial { ad.present(from: vc) } else { loadInterstitial() }
    }

    func showRewarded(from vc: UIViewController, placement: String?, reward: @escaping (String, Int) -> Void) {
        guard let ad = rewarded else { loadRewarded(); return }
        rewardCallback = reward
        ad.present(from: vc) {
            let r = ad.adReward
            reward(r.type, r.amount.intValue)
        }
    }

    func showAppOpen(from vc: UIViewController) {
        if let ad = appOpen { ad.present(from: vc) } else { loadAppOpen() }
    }

    // Reload after dismissal.
    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        if ad is InterstitialAd { interstitial = nil; loadInterstitial() }
        if ad is RewardedAd     { rewarded = nil;     loadRewarded() }
        if ad is AppOpenAd      { appOpen = nil;      loadAppOpen() }
    }
    func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        NSLog("[AdMob] present failed: \(error)")
    }
}
#endif

File: android/app/src/admob/android/AdMobAdsManager.kt
package co.median.android.ads

import android.app.Activity
import android.app.Application
import android.util.Log
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.appopen.AppOpenAd
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback

class AdMobAdsManager : AdsManager {
    private val tag = "AdMobAdsManager"
    private var app: Application? = null
    private var interstitial: InterstitialAd? = null
    private var rewarded: RewardedAd? = null
    private var appOpen: AppOpenAd? = null

    override fun init(app: Application) {
        this.app = app
        MobileAds.initialize(app) {
            Log.i(tag, "AdMob initialised")
            preload()
        }
    }

    override fun preload() {
        loadInterstitial(); loadRewarded(); loadAppOpen()
    }

    private fun loadInterstitial() {
        val ctx = app ?: return
        val id = AdsConfigHolder.config.admobInterstitial.ifBlank { TEST_INTERSTITIAL }
        InterstitialAd.load(ctx, id, AdRequest.Builder().build(), object : InterstitialAdLoadCallback() {
            override fun onAdLoaded(ad: InterstitialAd) { interstitial = ad }
            override fun onAdFailedToLoad(e: LoadAdError) { interstitial = null; Log.w(tag, "interstitial load failed: $e") }
        })
    }

    private fun loadRewarded() {
        val ctx = app ?: return
        val id = AdsConfigHolder.config.admobRewarded.ifBlank { TEST_REWARDED }
        RewardedAd.load(ctx, id, AdRequest.Builder().build(), object : RewardedAdLoadCallback() {
            override fun onAdLoaded(ad: RewardedAd) { rewarded = ad }
            override fun onAdFailedToLoad(e: LoadAdError) { rewarded = null; Log.w(tag, "rewarded load failed: $e") }
        })
    }

    private fun loadAppOpen() {
        val ctx = app ?: return
        val id = AdsConfigHolder.config.admobAppOpen.ifBlank { TEST_APP_OPEN }
        AppOpenAd.load(ctx, id, AdRequest.Builder().build(), object : AppOpenAd.AppOpenAdLoadCallback() {
            override fun onAdLoaded(ad: AppOpenAd) { appOpen = ad }
            override fun onAdFailedToLoad(e: LoadAdError) { appOpen = null; Log.w(tag, "appOpen load failed: $e") }
        })
    }

    override fun showInterstitial(activity: Activity, onClosed: (() -> Unit)?) {
        val ad = interstitial
        if (ad == null) { loadInterstitial(); onClosed?.invoke(); return }
        ad.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdDismissedFullScreenContent() { interstitial = null; loadInterstitial(); onClosed?.invoke() }
            override fun onAdFailedToShowFullScreenContent(e: AdError) { interstitial = null; loadInterstitial(); onClosed?.invoke() }
        }
        ad.show(activity)
    }

    override fun showRewarded(activity: Activity, placement: String?, onReward: (String, Int) -> Unit, onClosed: (() -> Unit)?) {
        val ad = rewarded
        if (ad == null) { loadRewarded(); onClosed?.invoke(); return }
        ad.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdDismissedFullScreenContent() { rewarded = null; loadRewarded(); onClosed?.invoke() }
            override fun onAdFailedToShowFullScreenContent(e: AdError) { rewarded = null; loadRewarded(); onClosed?.invoke() }
        }
        ad.show(activity) { reward -> onReward(reward.type ?: "", reward.amount) }
    }

    override fun showAppOpen(activity: Activity) {
        val ad = appOpen ?: run { loadAppOpen(); return }
        ad.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdDismissedFullScreenContent() { appOpen = null; loadAppOpen() }
            override fun onAdFailedToShowFullScreenContent(e: AdError) { appOpen = null; loadAppOpen() }
        }
        ad.show(activity)
    }

    companion object {
        // Google-provided test unit IDs (safe default).
        const val TEST_INTERSTITIAL = "ca-app-pub-3940256099942544/1033173712"
        const val TEST_REWARDED = "ca-app-pub-3940256099942544/5224354917"
        const val TEST_APP_OPEN = "ca-app-pub-3940256099942544/9257395921"
    }
}